lib/net/http/header.rb

```diff
@@ -806,56 +806,107 @@ def set_form_data(params, sep = '&')
 
   alias form_data= set_form_data
 
-  # Set an HTML form data set.
-  # +params+ :: The form data to set, which should be an enumerable.
-  #             See below for more details.
-  # +enctype+ :: The content type to use to encode the form submission,
-  #              which should be application/x-www-form-urlencoded or
-  #              multipart/form-data.
-  # +formopt+ :: An options hash, supporting the following options:
-  #              :boundary :: The boundary of the multipart message. If
-  #                           not given, a random boundary will be used.
-  #              :charset :: The charset of the form submission. All
-  #                          field names and values of non-file fields
-  #                          should be encoded with this charset.
-  #
-  # Each item of params should respond to +each+ and yield 2-3 arguments,
-  # or an array of 2-3 elements. The arguments yielded should be:
-  #
-  # - The name of the field.
-  # - The value of the field, it should be a String or a File or IO-like.
-  # - An options hash, supporting the following options
-  #   (used only for file uploads); entries:
+  # Stores form data to be used in a +POST+ or +PUT+ request.
+  #
+  # The form data given in +params+ consists of zero or more fields;
+  # each field is:
+  #
+  # - A scalar value.
+  # - A name/value pair.
+  # - An IO stream opened for reading.
+  #
+  # Argument +params+ should be an
+  # {Enumerable}[https://docs.ruby-lang.org/en/master/Enumerable.html#module-Enumerable-label-Enumerable+in+Ruby+Classes]
+  # (method <tt>params.map</tt> will be called),
+  # and is often an array or hash.
+  #
+  # First, we set up a request:
+  #
+  #   _uri = uri.dup
+  #   _uri.path ='/posts'
+  #   req = Net::HTTP::Post.new(_uri)
+  #
+  # <b>Argument +params+ As an Array</b>
+  #
+  # When +params+ is an array,
+  # each of its elements is a subarray that defines a field;
+  # the subarray may contain:
+  #
+  # - One string:
+  #
+  #     req.set_form([['foo'], ['bar'], ['baz']])
+  #     # String to be included in request body: 'foo&bar&baz'
+  #
+  # - Two strings:
+  #
+  #     req.set_form([%w[foo 0], %w[bar 1], %w[baz 2]])
+  #     # String to be include in request body: 'foo=0&bar=1&baz=2'
+  #
+  # - A string (ignored) and an IO stream opened for reading:
```

Comment:
1. This only applies for `multipart/form-data` and not `application/x-www-form-urlencoded`. For `application/x-www-form-urlencoded` it uses `URI.encode_www_form` which doesn't properly encode the IO stream:

```ruby
URI.encode_www_form([["foo", StringIO.new("foobar")]])
# => "foo=%23%3CStringIO%3A0x00000001060bdd88%3E"
```

  Whereas `'multipart/form-data'` does use the IO:

```ruby
io = StringIO.new
req.send(:encode_multipart_form_data, io, [["foo", StringIO.new("foobar")]], {})
puts io.string
# =>
# --KCyJx5jGk4z7PliS1mddX-fzCUfusQhEdg_EBzRp2SFxg17wzWP5Cw
# Content-Disposition: form-data; name="foo"
#
# foobar
# --KCyJx5jGk4z7PliS1mddX-fzCUfusQhEdg_EBzRp2SFxg17wzWP5Cw--
```

2. It does not ignore the first string. It uses it for the `name` field.

Reply:
1.  I've removed the "string to be included" remarks (four places), which I hope fixes this.
2. "string (ignored)" replaced by "string name" (two places).